### PR TITLE
fix(rds): advertise reachable endpoints

### PR DESCRIPTION
**Pull request by nineinchnick**
## Summary

RDS API responses advertised the proxy container port even when Docker published it on a different random host port. This made returned endpoints unreachable from hosts using VM-backed Docker runtimes.

This change lets callers configure the externally reachable endpoint host and resolves each RDS proxy published host port while keeping the internal listener port unchanged.

## Type of change

- [x] Bug fix (`fix:`)
- [ ] New feature (`feat:`)
- [ ] Breaking change (`feat!:` or `fix!:`)
- [ ] Docs / chore

## AWS Compatibility

The incorrect behavior was limited to the advertised RDS endpoint host and port when Floci runs in Docker with dynamic port publishing. AWS request and response wire shapes are unchanged.

Validated endpoint translation and Docker binding resolution with unit tests. The full Floci test suite passes. No integration test was added because successful-resolution caching is internal resolver behavior covered by the focused unit test.

## Checklist

- [x] `./mvnw test` passes locally
- [ ] New or updated integration test added
- [x] Commit messages follow [Conventional Commits](https://www.conventionalcommits.org/)

### CODE REVIEW

File: src/main/java/io/github/hectorvent/floci/services/rds/RdsService.java [hunk 1]
```diff
@@ -1092,6 +1100,19 @@ private void releaseProxyPort(int port) {
         usedPorts.remove(port);
     }
 
+    private DbEndpoint proxyEndpoint(int proxyPort) {
+        Optional<String> endpointHost = config.services().rds().endpointHost()
+                .filter(host -> !host.isBlank());
```
[the remark this answers is not in the document]
Comment (nineinchnick) [reply src/main/java/io/github/hectorvent/floci/services/rds/RdsService.java:1105]: MicroProfile Config's generated implementation honors the non-null Optional contract, and the existing RDS tests exercise unstubbed endpointHost() calls successfully through Mockito's Optional default. Treating null as absent would add production handling for an invalid config implementation, so I am leaving this unchanged.

File: src/main/java/io/github/hectorvent/floci/core/common/docker/CurrentContainerNetworkResolver.java [hunk 2]
```diff
@@ -45,6 +48,40 @@ public Optional<String> resolveContainerIp() {
...
+
+            Ports.Binding[] bindings = ports.getBindings().get(ExposedPort.tcp(containerPort));
+            if (bindings == null || bindings.length == 0 || bindings[0] == null) {
+                return OptionalInt.empty();
+            }
+
+            String hostPort = bindings[0].getHostPortSpec();
+            return hostPort == null || hostPort.isBlank()
+                    ? OptionalInt.empty()
+                    : OptionalInt.of(Integer.parseInt(hostPort));
```
[the remark this answers is not in the document]
Comment (nineinchnick) [reply src/main/java/io/github/hectorvent/floci/core/common/docker/CurrentContainerNetworkResolver.java]: Updated resolvePublishedPort to scan all bindings and return the first non-null binding with a non-blank host port. The regression test now places a null binding before the valid published-port mapping.

File: src/main/java/io/github/hectorvent/floci/config/EmulatorConfig.java [hunk 3]
```diff
@@ -862,6 +862,9 @@ interface RdsServiceConfig {
         @WithDefault("mariadb:11")
         String defaultMariadbImage();
 
+        /** Hostname advertised for RDS endpoints. Uses published Docker ports when configured. */
+        Optional<String> endpointHost();
```
[the remark this answers is not in the document]
Comment (nineinchnick) [reply src/main/java/io/github/hectorvent/floci/config/EmulatorConfig.java:866]: Documented endpoint-host in the main application.yml, advanced YAML reference, environment-variable reference, and RDS service guide. The guide now explains dynamic published-port translation and the compatible unset behavior.

File: src/main/java/io/github/hectorvent/floci/core/common/docker/CurrentContainerNetworkResolver.java [hunk 4]
```diff
@@ -45,6 +48,46 @@ public Optional<String> resolveContainerIp() {
         return resolve().map(CurrentContainerNetwork::ipAddress);
     }
 
+    public OptionalInt resolvePublishedPort(int containerPort) {
```
Comment (pgermosen) [inline src/main/java/io/github/hectorvent/floci/core/common/docker/CurrentContainerNetworkResolver.java:53]: Non-blocking: unlike `resolve()` just above (which caches `cachedNetwork` after the first Docker inspection), this does a fresh `inspectContainerCmd` on every call. Since `RdsService.proxyEndpoint()` calls this on both `createDbInstance`/`createDbCluster` and again on `startInstance`/`startCluster`, that's at least two live Docker API round-trips per resource — harmless today since the Floci container's own port-publish mapping is static for its whole lifetime (so every call returns the same answer), but worth caching the same way `resolve()` does if this ends up on a hot path.

**Review by pgermosen — COMMENTED** [review]
. The functional change is correct and well-tested; the issues are all process/hygiene ones.

**PR title doesn't follow Conventional Commits** — confirmed. `CONTRIBUTING.md` states: "**The PR title is validated automatically by CI** and must follow this format, since it becomes the squash-merge commit message that semantic-release reads." The title here, "Advertise reachable RDS endpoints", has no `<type>[scope]:` prefix — it's exactly the `Add PartiQL support # missing type` shape the doc's own "Invalid examples" section calls out. It should be renamed to match the PR's own (compliant) commit message: `fix(rds): advertise reachable endpoints`.

**The "validated automatically by CI" claim is itself inaccurate, which is why this slipped through green** — I checked `.github/workflows/conventional-commits.yml` directly: it lints `git log --no-merges --format="%s" "$BASE_SHA".."$HEAD_SHA"`, i.e. the actual *commit* messages, never `github.event.pull_request.title`. I grepped every workflow in `.github/workflows/` for any reference to the PR title (`title`, `semantic-pull-request`, etc.) and found none. This PR's single commit (`fix(rds): advertise reachable endpoints`) is compliant, so `Commits follow Conventional Commits` correctly shows green — but that green check doesn't actually mean the title (what becomes the squash commit message) is compliant, contrary to what the doc promises. Confirmed via `gh api repos/floci-io/floci --jq '{squash_merge_commit_title}'` that the repo has `squash_merge_commit_title: "COMMIT_OR_PR_TITLE"` — GitHub's documented behavior for that setting is: for a single-commit PR (this one), the squash box defaults to the *commit* message, not the PR title, so merging via the GitHub UI without hand-editing the box would likely still produce a compliant squash commit here. But that's incidental to this PR having exactly one commit, not something the repo's tooling actually guarantees — a multi-commit PR with a non-conventional title would default to the (non-compliant) PR title instead. Worth renaming the title regardless, since maintainers reasonably rely on "CI is green" as sufficient signal.

**PR description doesn't follow `.github/pull_request_template.md`** — no "Type of change" checklist, no "AWS Compatibility" section, no "Checklist" section; replaced with custom `## Why` / `## Approach` headers instead. Not a functional problem, but worth asking for since every other recent PR I've reviewed this session has followed the template, including the AWS-compatibility note that's genuinely relevant here (this is a non-AWS-facing operational fix, worth stating explicitly rather than leaving implicit).

On the code itself: traced `RdsService.proxyEndpoint()` end to end — `FLOCI_SERVICES_RDS_ENDPOINT_HOST` unset falls through to the pre-existing `proxyEndpointHost()`/raw-`proxyPort` behavior unchanged; set, it resolves the actual published host port via the new `CurrentContainerNetworkResolver.resolvePublishedPort()`, falling back to the raw `proxyPort` when not running in a container or when Docker inspection fails — sensible degradation in every branch. `resolvePublishedPort`'s `ExposedPort.tcp(containerPort)` lookup and null-binding handling are correct, and the new `resolvePublishedPort_skipsNullBinding` test genuinely exercises the null-array-slot edge case rather than just the happy path. Ran the full `Rds*Test` + `CurrentContainerNetworkResolverTest` suite myself: 159/159, no regressions. `make docs-check` exits clean (no hand-edits inside the `floci:actions` markers — the new docs are all outside them, in prose/config-reference sections). No CODEOWNERS-sensitive paths touched. No AI trailers.

File: src/main/java/io/github/hectorvent/floci/core/common/docker/CurrentContainerNetworkResolver.java (hunk 4 above)
Comment (nineinchnick) [reply src/main/java/io/github/hectorvent/floci/core/common/docker/CurrentContainerNetworkResolver.java:53]: Cached successful published-port resolutions per container port. Empty results and Docker inspection failures remain uncached so transient failures are retried.

**Review by pgermosen — APPROVED** [review]
Re-checked after the latest commit. Everything from the earlier rounds is genuinely fixed, not just replied-to:

- Title is now Conventional-Commits compliant, and the PR body follows the full template.
- `resolvePublishedPort` now caches only successful port resolutions per container port (confirmed in `CurrentContainerNetworkResolver.java`), leaving failures uncached so a transient Docker error gets retried rather than sticking — a better fix than what I'd suggested.
- The multi-binding scan (Copilot's finding) is in place and correctly grounded in Docker's own API contract: checked directly against moby/moby's `swagger.yaml` — `PortMap`'s values are explicitly `type: "array"`, keyed `"<port>/<protocol>"`, so a port can genuinely have more than one `PortBinding` entry (dual-stack IPv4/IPv6 publishing being the common case). Only checking the first entry really could miss the usable one; scanning all of them and skipping nulls is the correct fix, not defensive overkill.

Approving.
